doc/design/model_zoo.md

```diff
@@ -0,0 +1,174 @@
+
+1. Hyperparameters
+   1. **Model hyperparameters** are some key-value pairs specified in the WITH clause. They are arguments to the constructor of the model definition.
+   1. **Training hyperparameters** appear in WITH, as model constructors do. They affect the training process.
+   1. **Data conversion hyperparameters** appear in COLUMN and optionally LABEL. For example, the scale range and bucket size in the first example.
+
+1. A **trained model** consists of all the above concepts and the estimated model parameters.
+
+## The Design
+
+### Versioning and Releasing
+
+A key to this design is the representation of the above concepts.  It is not a straightforward solution. For example, in prior work and previous discussions, engineers proposed to represent each model definition by a Python source code file hosted on an FTP service and downloaded when some users file a model training statement. This intuitive proposal could lead to inconsistencies. Suppose that an analyst trained a model using the definition in `my_dnn_regressor.py` and got `my_first_model`; soon after that, a data scientist changed `my_dnn_regressor.py` and the analyst re-trained the model into `my_second_model` with slightly modified hyperparameter settings. The analyst might expect that both models share the same definition; however, they don't, and worse than that, there is no mechanism to remind the change of the model definition to the analyst.
+
+Such design flaw roots from the ignoring of the fact that model definitions are code that has versions. Another important aspect is that we often build the source code into some release-ready form.  Once we noticed these facts, we can use version management tools and Git and release engineering tools like Docker.  Here follows our proposal.
+
+1. A collection of model definitions is a Git repository of source files. 
+1. To describe dependencies, we require a Dockerfile at the root directory of the repository.
+1. To release a repository, we checkout the specific version and run `docker run` with the Dockerfile.
+1. Each release has a globally unique ID (GUID) represented by the hash ID of the Docker image.
+1. The Docker image must support a protocol of command-line arguments so that SQLFlow can pass in required information like hyperparameters from the `docker run` command.
+
+### Submitter Programs
+
+For a SELECT program using the SQLFlow syntax extension, the SQLFlow server converts it into a *submitter* program, usually in Python.
+
+Consider the training statement at the beginning of this document. Suppose that the Python class `MyDNNRegressor` is in source file `my_dnn_regressor.py` in the repository https://github.com/a_data_scientist/regressors, which builds into Docker image `a_data_scientist/regressors`.  Let us also assume that the submitter program is `/var/sqlflow/submitters/an_analyst/my_first_model.py`.  SQLFlow can run the submitter using the Docker image as follows:
```

Comment:
Please be clear `/var/sqlflow/submitters/an_analyst/my_first_model.py` is generated by SQLFlow. I was wondering what is `my_first_model.py` until I read the conversation at 

> The SQLFlow server invokes codegen_elasticdl.go to generates the submitter program my_first_model.py.